metis-repository/src/main/java/eu/europeana/metis/repository/rest/HttpHarvestController.java

```diff
@@ -0,0 +1,86 @@
+import java.io.BufferedOutputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.ObjectOutputStream;
+import java.util.List;
+import java.util.zip.ZipEntry;
+import java.util.zip.ZipOutputStream;
+import javax.servlet.http.HttpServletResponse;
+import org.apache.commons.compress.utils.IOUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.http.HttpStatus;
+import org.springframework.http.MediaType;
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.PathVariable;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.ResponseBody;
+import org.springframework.web.bind.annotation.ResponseStatus;
+import org.springframework.web.bind.annotation.RestController;
+
+/**
+ * Controller for HTTP harvesting.
+ */
+@RestController
+@RequestMapping("http")
+@Tags(@Tag(name = HttpHarvestController.CONTROLLER_TAG_NAME,
+    description = "Controller providing access to HTTP (zip) harvesting functionality."))
+@Api(tags = HttpHarvestController.CONTROLLER_TAG_NAME)
+public class HttpHarvestController {
+
+  public static final String CONTROLLER_TAG_NAME = "HttpHarvestController";
+  private static final Logger LOGGER = LoggerFactory.getLogger(HttpHarvestController.class);
+
+  private RecordDao recordDao;
+
+  @Autowired
+  void setRecordDao(RecordDao recordDao) {
+    this.recordDao = recordDao;
+  }
+
+  /**
+   * It creates a zip of records that belong to dataset with datasetId
+   * @return A zip of records
+   */
+  @GetMapping(value = RestEndpoints.GET_RECORDS_DATABASE, produces = "application/zip")
+  @ResponseStatus(HttpStatus.OK)
+  @ResponseBody
+  public byte[] getDatasetRecords(@PathVariable("dataset") String dataset,
+      HttpServletResponse response) throws IOException {
+
+    List<Record> recordList = recordDao.getAllRecordsFromDataset(dataset);
+
+    //setting headers
+    response.addHeader("Content-Disposition", "attachment; filename=\"result.zip\"");
+
+    ByteArrayOutputStream byteArrayOutputStream = new ByteArrayOutputStream();
+    BufferedOutputStream bufferedOutputStream = new BufferedOutputStream(byteArrayOutputStream);
+    ZipOutputStream zipOutputStream = new ZipOutputStream(bufferedOutputStream);
+
+    for (Record record : recordList) {
+      zipOutputStream.putNextEntry(new ZipEntry(record.getRecordId() + ".xml"));
+      ObjectOutputStream objectOutputStream = new ObjectOutputStream(zipOutputStream);
+      objectOutputStream.writeObject(record);
+      objectOutputStream.flush();
+      objectOutputStream.close();
+      zipOutputStream.closeEntry();
+    }
+
+    zipOutputStream.finish();
+    zipOutputStream.flush();
+    IOUtils.closeQuietly(zipOutputStream);
+    IOUtils.closeQuietly(bufferedOutputStream);
+    IOUtils.closeQuietly(byteArrayOutputStream);
```

Comment:
I think if there is a problem, it should report it (in the form of a 500 status code). So that the receiver knows that there is a problem.

Reply:
I changed the close and added a try catch block

Reply:
I ended up making some changes to the code. Making sure we use spring code only (no underlying java response object), returning a 500 code in case of issues, and using the try with closable scheme for the zip output stream. Have a look when you're back.